File: backend/app/routers/backtest.py
from fastapi import APIRouter, Query

from app.schemas.backtest import (
    BacktestResult,
    MonteCarloResult,
    PortfolioBacktestResult,
    StrategyCompareResult,
    TsmomResult,
    WalkForwardResult,
)
from app.services import backtest_service

router = APIRouter(prefix="/api/backtest", tags=["backtest"])


@router.get("/ma-cross", response_model=BacktestResult)
def ma_cross(
    market: str = Query("KRW-BTC"),
    fast: int   = Query(5,   ge=2, le=50),
    slow: int   = Query(20,  ge=5, le=200),
    count: int  = Query(200, ge=60, le=500),
    fee_bps: float = Query(5.0, ge=0, le=100, description="편도 거래비용(bps, 1bps=0.01%)"),
    target_vol: float = Query(0.0, ge=0, le=2, description="변동성 타게팅 목표(연율, 0=올인)"),
):
    return backtest_service.run_ma_cross(market, fast, slow, count, fee_bps, target_vol)


@router.get("/rsi", response_model=BacktestResult)
def rsi_strategy(
    market: str     = Query("KRW-BTC"),
    period: int     = Query(14,   ge=5,  le=30),
    oversold: float = Query(30.0, ge=10, le=45),
    overbought: float = Query(70.0, ge=55, le=90),
    count: int      = Query(200,  ge=60, le=500),
    fee_bps: float  = Query(5.0, ge=0, le=100, description="편도 거래비용(bps, 1bps=0.01%)"),
    target_vol: float = Query(0.0, ge=0, le=2, description="변동성 타게팅 목표(연율, 0=올인)"),
):
    return backtest_service.run_rsi_strategy(market, period, oversold, overbought, count, fee_bps, target_vol)


@router.get("/compare", response_model=StrategyCompareResult)
def compare(
    market: str = Query("KRW-BTC"),
    count: int  = Query(200, ge=60, le=500),
    fee_bps: float = Query(5.0, ge=0, le=100),
):
    return backtest_service.run_compare(market, count, fee_bps)


@router.get("/walk-forward", response_model=WalkForwardResult)
def walk_forward(
    market: str = Query("KRW-BTC"),
    count: int  = Query(300, ge=120, le=500),
    n_splits: int = Query(4, ge=2, le=8),
    fee_bps: float = Query(5.0, ge=0, le=100),
):
    return backtest_service.run_walk_forward(market, count, n_splits, fee_bps)


@router.get("/tsmom", response_model=TsmomResult)
def tsmom(
    top: int = Query(30, ge=10, le=100, description="유니버스 종목 수(거래대금 상위)"),
    lookback: int = Query(60, ge=10, le=180, description="추세 판단 룩백(일)"),
    holding: int = Query(5, ge=1, le=30, description="리밸런스 주기(일)"),
    count: int = Query(200, ge=120, le=500),
    fee_bps: float = Query(5.0, ge=0, le=100),
):
    return backtest_service.run_tsmom(top, lookback, holding, count, fee_bps)


@router.get("/montecarlo", response_model=MonteCarloResult)
def montecarlo(
    market: str = Query("KRW-BTC"),
    horizon: int = Query(30, ge=5, le=120, description="시뮬레이션 일수"),
    n_paths: int = Query(1000, ge=200, le=5000, description="경로 수"),
    count: int = Query(180, ge=60, le=500, description="과거 수익률 추정 윈도우(일봉)"),
):
    return backtest_service.run_montecarlo(market, horizon, n_paths, count)


@router.get("/portfolio", response_model=PortfolioBacktestResult)
def portfolio(
    markets: str = Query(..., description="쉼표 구분 마켓 코드 (예: KRW-BTC,KRW-ETH)"),
    weights: str | None = Query(None, description="쉼표 구분 비중 (markets와 같은 개수, 생략 시 동일가중)"),
    count: int = Query(180, ge=30, le=500, description="일봉 기간"),
    rebalance_days: int = Query(0, ge=0, le=90, description="리밸런스 주기(일), 0=매수보유"),
    fee_bps: float = Query(5.0, ge=0, le=100, description="편도 거래비용(bps, 1bps=0.01%)"),
):
    codes = [m.strip().upper() for m in markets.split(",") if m.strip()][:10]
    w = None
    if weights:
        try:
            w = [float(x) for x in weights.split(",") if x.strip()]
        except ValueError:
            w = None
    return backtest_service.run_portfolio(codes, w, count, rebalance_days, fee_bps)

File: frontend/src/api/backtest.ts
import api from './client'
import type {
  BacktestResult, StrategyCompareResult, WalkForwardResult,
  MonteCarloResult, TsmomResult, PortfolioBacktestResult,
} from '../types'

// axios query params로 그대로 전달되는 평면 객체.
type QueryParams = Record<string, string | number>

export const runMaCross = (params: QueryParams): Promise<BacktestResult> =>
  api.get('/api/backtest/ma-cross', { params }).then(r => r.data)

export const runRsi = (params: QueryParams): Promise<BacktestResult> =>
  api.get('/api/backtest/rsi', { params }).then(r => r.data)

// 다중 전략 겹쳐 비교 (한 종목에 MA·RSI 동시)
export const runCompare = ({ market, count = 200, fee_bps = 5 }: { market: string; count?: number; fee_bps?: number }): Promise<StrategyCompareResult> =>
  api.get('/api/backtest/compare', { params: { market, count, fee_bps } }).then(r => r.data)

// 워크포워드 (in-sample 그리드서치 → out-of-sample)
export const runWalkForward = ({ market, count = 300, n_splits = 4, fee_bps = 5 }: { market: string; count?: number; n_splits?: number; fee_bps?: number }): Promise<WalkForwardResult> =>
  api.get('/api/backtest/walk-forward', { params: { market, count, n_splits, fee_bps } }).then(r => r.data)

// 몬테카를로 시뮬레이션 — 미래 가격 경로 N개 부트스트랩
export const runMonteCarlo = ({ market, horizon = 30, n_paths = 1000, count = 180 }: { market: string; horizon?: number; n_paths?: number; count?: number }): Promise<MonteCarloResult> =>
  api.get('/api/backtest/montecarlo', { params: { market, horizon, n_paths, count } }).then(r => r.data)

// 시계열 모멘텀(추세추종) + 변동성 타게팅
export const runTsmom = ({ top = 30, lookback = 60, holding = 5, count = 200, fee_bps = 5 }: { top?: number; lookback?: number; holding?: number; count?: number; fee_bps?: number }): Promise<TsmomResult> =>
  api.get('/api/backtest/tsmom', { params: { top, lookback, holding, count, fee_bps } }).then(r => r.data)

// 포트폴리오 백테스트 — markets/weights는 쉼표 문자열로 직렬화
export const runPortfolio = ({ markets, weights, count = 180, rebalance_days = 0, fee_bps = 5 }: { markets: string[]; weights?: number[]; count?: number; rebalance_days?: number; fee_bps?: number }): Promise<PortfolioBacktestResult> =>
  api.get('/api/backtest/portfolio', {
    params: {
      markets: markets.join(','),
      ...(weights ? { weights: weights.join(',') } : {}),
      count, rebalance_days, fee_bps,
    },
  }).then(r => r.data)
